### PR TITLE
feat: see all permissions required to run a command

### LINKED ISSUE #511: Ability to call a given command's required permission
### Summary

It just returns what's in `@commands.has_permission()`.

### What is the feature request for?

nextcord.ext.commands

### The Problem

It'd be cool to have this in a help command that allows you to do `!help [command]`.

### The Ideal Solution

Probably something like `Command.required_permission`.

### The Current Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by EmreTech**
## Summary

This PR allows the user to see all of the permissions required to run a command set via the decorators `commands.has_permissions`, `commands.has_guild_permissions`, `commands.bot_has_permissions`, `commands.bot_has_guild_permissions` and their `ext.application_checks` counterparts. Unlike #618, this PR sets the attributes on the initial call. Closes #511.

## Checklist

- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
  - [ ] I have run `task pyright` and fixed the relevant issues.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by Skelmis — CHANGES_REQUESTED** [review]
There is a better way to do this which is to only set the attribute on initial call and not every time the command is invoked which was why this PR was originally closed due to the work required to accomplish it.

This rework requires changing the decors to have more nesting to allow for setting on the initial invoke as well as still supporting the checks themselves

**Comment by EmreTech** [comment]
> There is a better way to do this which is to only set the attribute on initial call and not every time the command is invoked which was why this PR was originally closed due to the work required to accomplish it.
> 
> This rework requires changing the decors to have more nesting to allow for setting on the initial invoke as well as still supporting the checks themselves.
> 
> This was discussed between myself and arl somewhere, hence why the last PR got closed. Didn't realize it wasn't documented.
> 
> If enough people are happy with the current approach however I will change this review to approval

I think setting the attribute on the initial call would be a much better approach. I'll try to rewrite the PR to accomplish that.

**Review by Skelmis — CHANGES_REQUESTED** [review]
Might as well add this to `ext.application_checks` as well in this PR

**Review by Skelmis — CHANGES_REQUESTED** [review]
Overall I approve, however the callback setters overlap such that this code says the test slash command requires manage guild perms when it does not.

```py
import asyncio
import os

from nextcord.ext import commands

bot = commands.Bot()


@bot.event
async def on_ready():
    await asyncio.sleep(5)
    cmd = bot.get_application_commands()[0]
    print(cmd.required_permissions)


@bot.slash_command()
@commands.has_permissions(manage_guild=True)
async def test(ctx):
    await ctx.send("Test")

bot.run(os.environ["TOKEN"])
```

**Comment by Skelmis** [comment]
> > Overall I approve, however the callback setters overlap such that this code says the test slash command requires manage guild perms when it does not.
> > ```python
> > import asyncio
> > import os
> > 
> > from nextcord.ext import commands
> > 
> > bot = commands.Bot()
> > 
> > 
> > @bot.event
> > async def on_ready():
> >     await asyncio.sleep(5)
> >     cmd = bot.get_application_commands()[0]
> >     print(cmd.required_permissions)
> > 
> > 
> > @bot.slash_command()
> > @commands.has_permissions(manage_guild=True)
> > async def test(ctx):
> >     await ctx.send("Test")
> > 
> > bot.run(os.environ["TOKEN"])
> > ```
> 
> Can you elaborate? I don't understand what you mean by "requires manage guild perms when it does not" in this example.

Well the way to set perms for slash commands in nextcord is via `ext.application_checks`

**Comment by EmreTech** [comment]
> > > Overall I approve, however the callback setters overlap such that this code says the test slash command requires manage guild perms when it does not.
> > > ```python
> > > import asyncio
> > > import os
> > > 
> > > from nextcord.ext import commands
> > > 
> > > bot = commands.Bot()
> > > 
> > > 
> > > @bot.event
> > > async def on_ready():
> > >     await asyncio.sleep(5)
> > >     cmd = bot.get_application_commands()[0]
> > >     print(cmd.required_permissions)
> > > 
> > > 
> > > @bot.slash_command()
> > > @commands.has_permissions(manage_guild=True)
> > > async def test(ctx):
> > >     await ctx.send("Test")
> > > 
> > > bot.run(os.environ["TOKEN"])
> > > ```
> > 
> > 
> > Can you elaborate? I don't understand what you mean by "requires manage guild perms when it does not" in this example.
> 
> Well the way to set perms for slash commands in nextcord is via `ext.application_checks`

I tested your example code, and the checks from `ext.commands` do work on application commands, but not the other way around. Are you asking me to check the type of the command passed into the decorators?

**Review by Skelmis — APPROVED** [review]

File: nextcord/ext/application_checks/core.py [hunk 1]
```diff
@@ -414,6 +414,21 @@ def predicate(interaction: Interaction) -> bool:
     return check(predicate)
 
 
+def _permission_check_wrapper(
+    predicate: "ApplicationCheck", name: str, perms: Dict[str, bool]
```
Comment (teaishealthy) [inline nextcord/ext/application_checks/core.py:418]:
```suggestion
    predicate: ApplicationCheck, name: str, perms: Dict[str, bool]
```

**Review by teaishealthy — CHANGES_REQUESTED** [review]

**Review by teaishealthy — APPROVED** [review]

**Pull request merged**